File: utils/src/main/java/com/lab/azure/bootloader/utils/DeployHelper.java
package com.lab.azure.bootloader.utils;

import com.lab.azure.bootloader.utils.arguments.DeployArgument;
import com.microsoft.azure.storage.CloudStorageAccount;
import com.microsoft.azure.storage.StorageException;
import com.microsoft.azure.storage.table.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URISyntaxException;
import java.security.InvalidKeyException;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;


public class DeployHelper {

    public static void deploy(DeployArgument deployArgument) throws StorageException, URISyntaxException, InvalidKeyException, IOException {
        Logger logger = LoggerFactory.getLogger("deploy");
        logger.info("deploy {} start", deployArgument.getTask());
        Settings settings = Settings.loadCfg(deployArgument.getSettingFile());
        if (!settings.validate()) {
            logger.warn("settings error");
            System.exit(-1);
        }
        if (!Helper.checkKey(deployArgument.getTask())) {
            logger.warn("task key {} error", deployArgument.getTask());
            System.exit(-1);
        }
        CloudStorageAccount account = CloudStorageAccount.parse(settings.getAzureWebJobsStorage());
        CloudTableClient tableClient = account.createCloudTableClient();
        CloudTable jobHistoryTable = tableClient.getTableReference(settings.getJobHistoryTable());
        CloudTable deployTable = tableClient.getTableReference(settings.getDeployTable());
        if (!jobHistoryTable.exists()) {
            logger.warn("jobHistoryTable not exists");
            System.exit(-1);
        }
        if (!deployTable.exists()) {
            logger.warn("jobHistoryTable not exists");
            System.exit(-1);
        }

        JobEntity jobEntity;
        {
            String pfilter = TableQuery.generateFilterCondition("PartitionKey", TableQuery.QueryComparisons.EQUAL, deployArgument.getTask());
            String rfilter = TableQuery.generateFilterCondition("RowKey", TableQuery.QueryComparisons.EQUAL, deployArgument.getTask());
            String cfilter = TableQuery.combineFilters(pfilter, TableQuery.Operators.AND, rfilter);
            TableQuery<JobEntity> query = TableQuery.from(JobEntity.class).where(cfilter);
            Iterable<JobEntity> jobEntityIterable = jobHistoryTable.execute(query);
            List<JobEntity> jobEntities = StreamSupport.stream(jobEntityIterable.spliterator(), false).collect(Collectors.toList());
            if(jobEntities.size()!=1){
                logger.warn("job Entity not 1");
                System.exit(-1);
            }
            jobEntity = jobEntities.get(0);
        }
        {
            DynamicTableEntity deploy = new DynamicTableEntity(jobEntity.getVersion(), "latest");
            HashMap<String, EntityProperty> map = new HashMap<>();
            map.put("TaskId",new EntityProperty(jobEntity.getPartitionKey()));
            map.put("Threshold",new EntityProperty(jobEntity.getAmount()));
            map.put("UpdateAt",new EntityProperty(new Date()));
            deploy.setProperties(map);
            TableOperation insertOrMerge = TableOperation.insertOrMerge(deploy);
            deployTable.execute(insertOrMerge);
        }
        logger.info("deploy {} stop", deployArgument.getTask());

    }
}

File: src/AzureBootloaderCompiler.Utils/DeployHelper.cs
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Table;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using AzureBootloaderCompiler.Utils.Arguments;

namespace AzureBootloaderCompiler.Utils
{
    class DeployHelper
    {
        public static async Task DoDeployAsync(DeployArgument argument)
        {
            Console.WriteLine($"deploy {argument.Task} start");
            Settings settings = Settings.LoadCfg(argument.SettingFile);
            if (!CheckKey(argument.Task))
            {
                Console.WriteLine($"task key {argument.Task} error");
                Environment.Exit(-1);
            }
            var account = CloudStorageAccount.Parse(settings.AzureWebJobsStorage);
            var tableClient = account.CreateCloudTableClient();
            var jobHistoryTable = tableClient.GetTableReference(settings.JobHistoryTable);
            var deployTable = tableClient.GetTableReference(settings.DeployTable);

            if (!await jobHistoryTable.ExistsAsync())
            {
                Console.WriteLine("jobHistoryTable not exists");
                Environment.Exit(-1);
            }
            if (!await deployTable.ExistsAsync())
            {
                Console.WriteLine("jobHistoryTable not exists");
                Environment.Exit(-1);
            }

            JobEntity jobEntity;
            {
                var partitionKey = argument.Task;
                var rowKey = argument.Task;
                TableOperation retrieveOperation = TableOperation.Retrieve<JobEntity>(partitionKey, rowKey);
                TableResult result = await jobHistoryTable.ExecuteAsync(retrieveOperation);
                jobEntity = result.Result as JobEntity;

                if (jobEntity == null)
                {
                    Console.WriteLine("job Entity not 1");
                    Environment.Exit(-1);
                }
            }

            {
                DynamicTableEntity deploy = new DynamicTableEntity(jobEntity.Version, "latest");
                Dictionary<string, EntityProperty> pairs = new Dictionary<string, EntityProperty>();

                pairs["TaskId"] = new EntityProperty(jobEntity.PartitionKey);
                pairs["Threshold"] = new EntityProperty(jobEntity.Amount);
                pairs["UpdateAt"] = new EntityProperty(new DateTime());
                deploy.Properties = pairs;
                TableOperation insertOrMerge = TableOperation.InsertOrMerge(deploy);
                _ = deployTable.ExecuteAsync(insertOrMerge);
                Console.WriteLine($"deploy {argument.Task} stop");
            }
        }

        private static Boolean CheckKey(string keyString)
        {
            try
            {
                Guid uuid = Guid.Parse(keyString);
                return true;
            }
            catch (FormatException)
            {
                return false;
            }
        }
    }
}
